### PR TITLE
feat: add validateAccessJwt to cloudflare:workers

### PR DESCRIPTION
**Pull request by elithrar**
Adds a `validateAccessJwt()` function to the `cloudflare:workers` module that validates Cloudflare Access JWTs against team-specific JWKs, providing a built-in alternative for users to validate Access JWTs directly vs. glueing it all together themselves.

- Throws `AccessJwtError` with specific error codes (`ERR_JWT_MISSING`, `ERR_JWT_EXPIRED`, etc.) on validation failure - defaults to throwing exceptions that you must catch vs. some `isValidJWT()` type function that can be mishandled
- No external dependencies - uses WebCrypto APIs
- Retry logic for JWKS fetch (3 attempts, 5s backoff on 5xx errors)
- Isolate-level JWKS caching (1 hour TTL, with cache invalidation on key rotation)
- Team domain normalization (accepts both `"myteam"` and `"myteam.cloudflareaccess.com"`)
- 60s clock skew tolerance for expiration

#### Usage

```typescript
import { validateAccessJwt, AccessJwtError } from 'cloudflare:workers';

export default {
  async fetch(request, env) {
    try {
      // env.TEAM_DOMAIN is their CF1 Team to fetch the JWKSet from
      // env.AUDIENCE is the ID from their Access app (per the Access docs)
      const payload = await validateAccessJwt(request, env.TEAM_DOMAIN, env.AUDIENCE);
      return new Response(`Hello ${payload.email}!`);
    } catch (err) {
      if (err instanceof AccessJwtError) {
        return new Response(`Auth failed: ${err.code}`, { status: 403 });
      }
      throw err;
    }
  }
};
```

#### Error Codes

These will be added to the public docs under https://developers.cloudflare.com/cloudflare-one/access-controls/applications/http-apps/authorization-cookie/validating-json/ + new Workers `/examples` as well

| Code | Description |
|------|-------------|
| `ERR_JWT_MISSING` | No `cf-access-jwt-assertion` header |
| `ERR_JWT_MALFORMED` | Invalid JWT structure or unsupported algorithm |
| `ERR_JWT_INVALID_SIGNATURE` | Signature verification failed |
| `ERR_JWT_EXPIRED` | Token `exp` claim is in the past |
| `ERR_JWT_NOT_YET_VALID` | Token `nbf` claim is in the future |
| `ERR_JWT_AUDIENCE_MISMATCH` | Token `aud` doesn't match expected audience |
| `ERR_JWT_ISSUER_MISMATCH` | Token `iss` doesn't match team domain |
| `ERR_JWKS_FETCH_FAILED` | Failed to fetch JWKS after retries |
| `ERR_JWKS_NO_MATCHING_KEY` | No key with matching `kid` found |
| `ERR_INVALID_TEAM_DOMAIN` | Empty team domain provided |
| `ERR_INVALID_AUDIENCE` | Empty audience provided |

## Test Plan

- 20 test cases covering all error codes and edge cases
- Run: `just test '//src/cloudflare/internal/test/access-jwt:access-jwt-test@'`

### CODE REVIEW

**Comment by elithrar** [comment]
The tests (after building...) fail locally — unclear (to me) why:

```
$ cat /private/var/tmp/_bazel_matt/2a80ed57af32aa0a2e438bf225f1030d/execroot/_main/bazel-out/darwin_arm64-fastbuild/testlogs/src/cloudflare/internal/test/access-jwt/access-jwt-test@/test.log
Executing tests from //src/cloudflare/internal/test/access-jwt:access-jwt-test@
-----------------------------------------------------------------------------
kj/exception.c++:357: warning: llvm-symbolizer was not found. To symbolize stack traces, install it in your $PATH or set $LLVM_SYMBOLIZER to the location of the binary. When running tests under bazel, use `--test_env=LLVM_SYMBOLIZER=<path>`.
*** Received signal #11: Segmentation fault: 11
stack: 10b8c7907 10b8c6a3f 10b8c650b 107c8ea5f 107c8fb1b 1071b9fcf 1071b8b63 1071b8a17 1071b8983 1074fcdc3 1074fd943 106c8020b 106c8014f 106c8011b 106c7c79f 106c7c3f7 104fe02ab 104fe01f3 104fdf173 104fdec37 104fdeac7 104984d1b 104984b9b 104984e63 10477d91b 1048d8cb7 10bb5b4ab 10bb5b477 10bb52f87
```

File: src/cloudflare/internal/access-jwt.ts [hunk 1]
```diff
@@ -0,0 +1,423 @@
...
+ * Base64url decode to ArrayBuffer for signature verification.
+ */
+function base64UrlDecodeToBytes(input: string): ArrayBuffer {
+  const binary = base64UrlDecode(input);
+  const bytes = new Uint8Array(binary.length);
+  for (let i = 0; i < binary.length; i++) {
+    bytes[i] = binary.charCodeAt(i);
+  }
+  return bytes.buffer as ArrayBuffer;
+}
```
Comment (anonrig) [inline src/cloudflare/internal/access-jwt.ts]: What about `Uint8Array.fromBase64(base64).buffer`?

Ref: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint8Array/fromBase64

File: src/cloudflare/internal/access-jwt.ts [hunk 2]
```diff
@@ -0,0 +1,423 @@
...
+  // Add padding if needed
+  const padding = base64.length % 4;
+  if (padding === 2) {
+    base64 += '==';
+  } else if (padding === 3) {
+    base64 += '=';
+  }
+
+  return atob(base64);
+}
```
Comment (anonrig) [inline src/cloudflare/internal/access-jwt.ts]: atob is not recommended to be used. we now have uint8array.frombase64 etc.

File: src/cloudflare/internal/access-jwt.ts [hunk 3]
```diff
@@ -0,0 +1,423 @@
...
+    throw new AccessJwtError(
+      'ERR_JWT_MALFORMED',
+      `JWT must have 3 parts, got ${parts.length}`
+    );
+  }
+
+  const [headerB64, payloadB64, signatureB64] = parts;
+
+  try {
+    const headerJson = base64UrlDecode(headerB64!);
```
Comment (anonrig) [inline src/cloudflare/internal/access-jwt.ts]: Can we avoid having !

```suggestion
    const headerJson = base64UrlDecode(headerB64);
```

File: src/cloudflare/internal/access-jwt.ts [hunk 4]
```diff
@@ -0,0 +1,423 @@
...
+} {
+  const parts = token.split('.');
+  if (parts.length !== 3) {
+    throw new AccessJwtError(
+      'ERR_JWT_MALFORMED',
+      `JWT must have 3 parts, got ${parts.length}`
+    );
+  }
+
+  const [headerB64, payloadB64, signatureB64] = parts;
```
Comment (anonrig) [inline src/cloudflare/internal/access-jwt.ts]: With this change, you won't need exclamation marks

```suggestion
  const [headerB64, payloadB64, signatureB64] = parts as FixedLengthArray<3, string>;
```

Yes, we have FixedLengthArray type in this codebase :-))

File: src/cloudflare/internal/access-jwt.ts [hunk 5]
```diff
@@ -0,0 +1,423 @@
...
+    );
+  }
+}
+
+/**
+ * Sleep for the specified duration.
+ */
+function sleep(ms: number): Promise<void> {
+  return new Promise((resolve) => setTimeout(resolve, ms));
+}
```
Comment (anonrig) [inline src/cloudflare/internal/access-jwt.ts:186]: we have `scheduler.wait()` or `node:timers/promises`

File: src/cloudflare/internal/access-jwt.ts [hunk 6]
```diff
@@ -0,0 +1,423 @@
...
+        );
+      }
+
+      // 5xx errors: continue to retry
+      lastError = new Error(`HTTP ${res.status} ${res.statusText}`);
+    } catch (e) {
+      if (e instanceof AccessJwtError) {
+        throw e;
+      }
+      lastError = e instanceof Error ? e : new Error('Unknown fetch error');
```
Comment (anonrig) [inline src/cloudflare/internal/access-jwt.ts]:
```suggestion
      lastError = Error.isError(e) ? e : new Error('Unknown fetch error', { cause: e });
```

File: src/cloudflare/internal/access-jwt.ts [hunk 7]
```diff
@@ -0,0 +1,423 @@
...
+async function importKey(jwk: JwkKey): Promise<CryptoKey> {
+  return crypto.subtle.importKey(
+    'jwk',
+    {
+      kty: jwk.kty,
+      n: jwk.n,
+      e: jwk.e,
+      alg: jwk.alg,
+      use: jwk.use,
+    },
```
Comment (anonrig) [inline src/cloudflare/internal/access-jwt.ts]:
```suggestion
    'jwk',
    jwk,
```

File: src/cloudflare/internal/access-jwt.ts (hunk 1 above)
Comment (elithrar) [reply src/cloudflare/internal/access-jwt.ts]: Way better — missed that this supports the base64url alphabet!

File: src/cloudflare/internal/test/access-jwt/access-jwt-mock.js [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+  if (typeof data === 'string') {
+    base64 = btoa(data);
+  } else {
+    // ArrayBuffer or Uint8Array
+    const bytes = new Uint8Array(data);
+    let binary = '';
+    for (let i = 0; i < bytes.length; i++) {
+      binary += String.fromCharCode(bytes[i]);
+    }
+    base64 = btoa(binary);
```
Comment (anonrig) [inline src/cloudflare/internal/test/access-jwt/access-jwt-mock.js]: We can also use Uint8array.toBase64

Comment (elithrar) [reply src/cloudflare/internal/test/access-jwt/access-jwt-mock.js]: got it - having an issue seeing that type (the build env is tricky to set up for a newcomer - can't even fetch the bazel version - 404s)

File: src/cloudflare/internal/access-jwt.ts (hunk 1 above)
Comment (anonrig) [reply src/cloudflare/internal/access-jwt.ts]: Yes! (Courtesy of @lemire and his work on simdutf!)

File: src/cloudflare/internal/test/access-jwt/access-jwt-mock.js [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+}
+
+/**
+ * Get or create a key pair for a team domain.
+ */
+async function getKeyPair(teamDomain) {
+  if (!keyPairs.has(teamDomain)) {
+    keyPairs.set(teamDomain, await generateKeyPair());
+  }
+  return keyPairs.get(teamDomain);
```
Comment (anonrig) [inline src/cloudflare/internal/test/access-jwt/access-jwt-mock.js]: Nit: This would be faster (not that it matters)

```suggestion
  const value = keyPairs.get(teamDomain);
  if (value == null) {
    keyPairs.set(teamDomain, await generateKeyPair());
  }
  return ;
```

**Review by anonrig — DISMISSED** [review]

File: src/cloudflare/internal/test/access-jwt/access-jwt-mock.js [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+function base64UrlEncode(data) {
+  let bytes;
+  if (typeof data === 'string') {
+    bytes = new TextEncoder().encode(data);
+  } else {
+    bytes = new Uint8Array(data);
+  }
+  // Use Uint8Array.prototype.toBase64 and convert to base64url
+  const base64 = bytes.toBase64();
+  return base64.replace(/\+/g, '-').replace(/\//g, '_').replace(/=/g, '');
```
Comment (mhart) [inline src/cloudflare/internal/test/access-jwt/access-jwt-mock.js]: Can be:
```suggestion
  return bytes.toBase64({ alphabet: 'base64url', omitPadding: true });
```

Comment (mhart) [reply src/cloudflare/internal/test/access-jwt/access-jwt-mock.js]: Oh this is just in the test, lol. But still, might help for AI SEO or future agent modifications 😉

File: src/cloudflare/internal/access-jwt.ts [hunk 11]
```diff
@@ -0,0 +1,428 @@
...
+  Uint8Array as unknown as Uint8ArrayConstructorWithBase64;
+
+/**
+ * Base64url decode to string (RFC 4648 §5).
+ * Converts URL-safe base64 to standard base64 and decodes to UTF-8 string.
+ */
+function base64UrlDecodeToString(input: string): string {
+  // Replace URL-safe characters with standard base64
+  const base64 = input.replace(/-/g, '+').replace(/_/g, '/');
+  const bytes = Uint8ArrayWithBase64.fromBase64(base64);
```
Comment (mhart) [inline src/cloudflare/internal/access-jwt.ts]: Any reason not to use the official method?
```suggestion
  const bytes = Uint8ArrayWithBase64.fromBase64(base64, { alphabet: 'base64url' });
```

File: src/cloudflare/internal/access-jwt.ts [hunk 12]
```diff
@@ -0,0 +1,428 @@
...
+  return new TextDecoder().decode(bytes);
```
Comment (mhart) [inline src/cloudflare/internal/access-jwt.ts]: Worth considering whether to just put a `const decoder = new TextDecoder()` at the module level and then just reuse that everywhere. Same with any `TextEncoder` you use

File: src/cloudflare/internal/access-jwt.ts [hunk 13]
```diff
@@ -0,0 +1,428 @@
...
+}
+
+/**
+ * Base64url decode to ArrayBuffer for signature verification.
+ */
+function base64UrlDecodeToBytes(input: string): ArrayBuffer {
+  // Replace URL-safe characters with standard base64
+  const base64 = input.replace(/-/g, '+').replace(/_/g, '/');
+  return Uint8ArrayWithBase64.fromBase64(base64).buffer as ArrayBuffer;
```
Comment (mhart) [inline src/cloudflare/internal/access-jwt.ts]:
```suggestion
  return Uint8ArrayWithBase64.fromBase64(base64, { alphabet: 'base64url' }).buffer as ArrayBuffer;
```

File: src/cloudflare/internal/test/access-jwt/access-jwt-mock.js [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+
+  const keyPair = await getKeyPair(teamDomain);
+
+  const jwks = {
+    keys: [keyPair.publicJwk],
+  };
+
+  return new Response(JSON.stringify(jwks), {
+    headers: { 'Content-Type': 'application/json' },
+  });
```
Comment (mhart) [inline src/cloudflare/internal/test/access-jwt/access-jwt-mock.js]:
```suggestion
  return Response.json(jwks);
```

File: src/cloudflare/internal/access-jwt.ts [hunk 15]
```diff
@@ -0,0 +1,428 @@
...
+
+/**
+ * Verify the JWT signature using the provided key.
+ */
+async function verifySignature(
+  key: CryptoKey,
+  signature: ArrayBuffer,
+  signedPart: string
+): Promise<boolean> {
+  const encoder = new TextEncoder();
```
Comment (mhart) [inline src/cloudflare/internal/access-jwt.ts]: Could be at the module level

File: src/cloudflare/internal/access-jwt.ts [hunk 16]
```diff
@@ -0,0 +1,428 @@
...
+  const trimmed = teamDomain.trim().toLowerCase();
+  if (trimmed.endsWith('.cloudflareaccess.com')) {
+    return trimmed;
+  }
+  return `${trimmed}.cloudflareaccess.com`;
+}
+
+// Uint8Array.fromBase64 is available at runtime but not in TS types yet
+type Uint8ArrayWithBase64 = Uint8Array & {
+  toBase64(): string;
```
Comment (mhart) [inline src/cloudflare/internal/access-jwt.ts]: From the [official TS source](https://github.com/dontwanttothink/TypeScript/blob/main/src/lib/esnext.typedarrays.d.ts):
```suggestion
  toBase64(
    options?: {
      alphabet?: "base64" | "base64url" | undefined;
      omitPadding?: boolean | undefined;
    },
  ): string;
```

File: src/cloudflare/internal/access-jwt.ts [hunk 17]
```diff
@@ -0,0 +1,428 @@
...
+};
+interface Uint8ArrayConstructorWithBase64 {
+  new (length: number): Uint8ArrayWithBase64;
+  fromBase64(base64: string): Uint8ArrayWithBase64;
```
Comment (mhart) [inline src/cloudflare/internal/access-jwt.ts]: From the [official TS source](https://github.com/dontwanttothink/TypeScript/blob/main/src/lib/esnext.typedarrays.d.ts):
```suggestion
  fromBase64(
    string: string,
    options?: {
      alphabet?: "base64" | "base64url" | undefined;
      lastChunkHandling?: "loose" | "strict" | "stop-before-partial" | undefined;
    },
  ): Uint8ArrayWithBase64;
```

**Review by mhart — APPROVED** [review]
Some minor suggestions, just to use some slightly more modern/idiomatic methods, but looks great overall

File: src/cloudflare/internal/access-jwt.ts (hunk 1 above)
Comment (elithrar) [reply src/cloudflare/internal/access-jwt.ts]: ack

File: src/cloudflare/internal/access-jwt.ts (hunk 2 above)
Comment (elithrar) [reply src/cloudflare/internal/access-jwt.ts]: ack

File: src/cloudflare/internal/access-jwt.ts (hunk 3 above)
Comment (elithrar) [reply src/cloudflare/internal/access-jwt.ts]: ack

File: src/cloudflare/internal/access-jwt.ts (hunk 4 above)
Comment (elithrar) [reply src/cloudflare/internal/access-jwt.ts]: ack

File: src/cloudflare/internal/access-jwt.ts (hunk 5 above)
Comment (elithrar) [reply src/cloudflare/internal/access-jwt.ts:186]: ack

File: src/cloudflare/internal/access-jwt.ts (hunk 6 above)
Comment (elithrar) [reply src/cloudflare/internal/access-jwt.ts]: ack

File: src/cloudflare/internal/access-jwt.ts (hunk 7 above)
Comment (elithrar) [reply src/cloudflare/internal/access-jwt.ts]: ack

File: src/cloudflare/internal/test/access-jwt/access-jwt-mock.js (hunk 8 above)
Comment (elithrar) [reply src/cloudflare/internal/test/access-jwt/access-jwt-mock.js]: ack

File: src/cloudflare/internal/test/access-jwt/access-jwt-mock.js (hunk 9 above)
Comment (elithrar) [reply src/cloudflare/internal/test/access-jwt/access-jwt-mock.js]: ack

File: src/cloudflare/internal/access-jwt.ts (hunk 12 above)
Comment (elithrar) [reply src/cloudflare/internal/access-jwt.ts]: ack

File: src/cloudflare/internal/access-jwt.ts (hunk 13 above)
Comment (elithrar) [reply src/cloudflare/internal/access-jwt.ts]: ack

File: src/cloudflare/internal/test/access-jwt/access-jwt-mock.js (hunk 14 above)
Comment (elithrar) [reply src/cloudflare/internal/test/access-jwt/access-jwt-mock.js]: ack

File: src/cloudflare/internal/access-jwt.ts (hunk 15 above)
Comment (elithrar) [reply src/cloudflare/internal/access-jwt.ts]: ack

File: src/cloudflare/internal/access-jwt.ts (hunk 16 above)
Comment (elithrar) [reply src/cloudflare/internal/access-jwt.ts]: ack

File: src/cloudflare/internal/access-jwt.ts (hunk 17 above)
Comment (elithrar) [reply src/cloudflare/internal/access-jwt.ts]: ack

File: src/cloudflare/internal/access-jwt.ts (hunk 7 above)
Comment (elithrar) [reply src/cloudflare/internal/access-jwt.ts]: ack

File: src/cloudflare/internal/test/access-jwt/access-jwt-mock.js (hunk 8 above)
Comment (elithrar) [reply src/cloudflare/internal/test/access-jwt/access-jwt-mock.js]: ack

File: src/cloudflare/internal/access-jwt.ts (hunk 13 above)
Comment (elithrar) [reply src/cloudflare/internal/access-jwt.ts]: ack

File: src/cloudflare/internal/access-jwt.ts (hunk 16 above)
Comment (elithrar) [reply src/cloudflare/internal/access-jwt.ts]: ack

**Comment by elithrar** [comment]
@mhart @anonrig I think I got this right — addressed all the comments:

* replaced atob/btoa with Uint8Array.fromBase64/toBase64 w/ base64url alphabet
* used FixedLengthArray<string, 3>
* switched from setTimeout to globalThis.scheduler.wait() for retry delays
* module-level TextEncoder/TextDecoder for reuse

File: src/cloudflare/internal/access-jwt.ts [hunk 18]
```diff
@@ -0,0 +1,459 @@
...
+  });
+  return textDecoder.decode(bytes);
+}
+
+/**
+ * Base64url decode to ArrayBuffer for signature verification.
+ */
+function base64UrlDecodeToBytes(input: string): ArrayBuffer {
+  return Uint8ArrayWithBase64.fromBase64(input, { alphabet: 'base64url' })
+    .buffer as ArrayBuffer;
```
Comment (jasnell) [inline src/cloudflare/internal/access-jwt.ts]: nit: feels a bit odd to have this as a separate utility function but not use it in `base64UrlDecodeToString` above.

Comment (elithrar) [reply src/cloudflare/internal/access-jwt.ts]: good call - I think we should just call `Uint8ArrayWithBase64.fromBase64(input, { alphabet: 'base64url' }).buffer as ArrayBuffer` directly. not worth a util function for something so small.

**Comment by anonrig** [comment]
Except the failing tests, the implementation looks really good.

File: src/cloudflare/internal/access-jwt.ts [hunk 19]
```diff
@@ -0,0 +1,449 @@
...
+ */
+function sleep(ms: number): Promise<void> {
+  return globalThis.scheduler.wait(ms);
+}
+
+/**
+ * Fetch JWKS from Cloudflare Access with retry logic for 5xx errors.
+ */
+async function fetchJwks(normalizedDomain: string): Promise<JwkSet> {
+  const url = `https://${normalizedDomain}/cdn-cgi/access/certs`;
```
Comment (anonrig) [inline src/cloudflare/internal/access-jwt.ts:192]: Nit: best way to do this is by doing:

```
const url = new URL(`https://cloudflare.com/cdn-cgi/access/certs`)
url.hostname = normalizedDomain;
```

Comment (jasnell) [reply src/cloudflare/internal/access-jwt.ts:192]: shouldn't that be `const url = new URL('https://cloudflare.com/cdn-cgi/access/certs`);` ?

**Comment by elithrar** [comment]
@anonrig unfortunately getting the tests to run locally in any reliable way is beyond me, and I can't grok why the tests can't see the `access-jwt` module:
```
FAIL: //src/cloudflare/internal/test/access-jwt:access-jwt-test@ (Exit 1) (see /home/runner/.cache/bazel/_bazel_runner/be54bd8ac890d1817c772d91eb8f6cb4/execroot/_main/bazel-out/aarch64-dbg/testlogs/src/cloudflare/internal/test/access-jwt/access-jwt-test@/test.log)
INFO: From Testing //src/cloudflare/internal/test/access-jwt:access-jwt-test@:
==================== Test output for //src/cloudflare/internal/test/access-jwt:access-jwt-test@:
service access-jwt-test: Uncaught Error: No such module "cloudflare-internal:access-jwt".
  imported from "worker" 
```

File: src/cloudflare/internal/access-jwt.ts (hunk 19 above)
Comment (elithrar) [reply src/cloudflare/internal/access-jwt.ts:192]: The domain has to include the team name: e.g. `https://opencode-devtools.cloudflareaccess.com/cdn-cgi/access/certs` - as the certs are scoped to the Access team/domain

Comment (anonrig) [reply src/cloudflare/internal/access-jwt.ts:192]: You can just do `url.hostname = "opencode-devtools.cloudflareaccess.com"` on the new URL() and it would be a lot safer

**Comment by anonrig** [comment]
> @anonrig unfortunately getting the tests to run locally in any reliable way is beyond me, and I can't grok why the tests can't see the `access-jwt` module:

@elithrar Sorry I just saw this comment. Unfortunately, internal modules are not available on test harness.

**Review by jroyal — APPROVED** [review]

File: src/cloudflare/internal/access-jwt.ts (hunk 19 above)
Comment (elithrar) [reply src/cloudflare/internal/access-jwt.ts:192]: I don't think I understand the comments here :)

* The certs are always at `https://${teamName}.cloudflareaccess.com/cdn-cgi/access/certs` - `teamName` is dynamic
* It's not static (to be extra clear)

We can make this `https://${teamName}.cloudflareaccess.com/cdn-cgi/access/certs` so that the string template doesn't need to cover the 'static' part of the domain but I don't see that as a net win?

**Comment by elithrar** [comment]
Changes to fix the test module import issue:

1. Removed `_clearJwksCache` from `access-jwt.ts` (dead code since tests can't access internal modules)
2. Removed the `cloudflare-internal:access-jwt` import from the test file
3. Each test now uses a unique team domain (e.g., `test-valid`, `test-expired`, `test-5xx-retry`) so cache entries don't conflict between tests

**Comment by elithrar** [comment]
For internal reviewers: https://share.opencode.cloudflare.dev/share/fqzhroMQ

**Comment by elithrar** [comment]
@anonrig might need your help to get this over the line from a toolchain PoV.

File: src/cloudflare/workers.ts [hunk 20]
```diff
@@ -8,6 +8,13 @@
 import entrypoints from 'cloudflare-internal:workers';
 import innerEnv from 'cloudflare-internal:env';
 
+export {
+  validateAccessJwt,
+  AccessJwtError,
+  type AccessJwtErrorCode,
+  type AccessJwtPayload,
+} from 'cloudflare-internal:access-jwt';
```
Comment (kentonv) [inline src/cloudflare/workers.ts:16]: Adding this to the `cloudflare:workers` module means that this JavaScript code will need to be parsed and compiled by every isolate that imports `cloudflare:workers`, whether or not they actually use the JWT functions. The `cloudflare:workers` module is relevant to almost everyone these days, but the JWT functions are relevant to a much smaller set of people. Also, there's a proposal to add JWT validation to the native APIs which would make these functions obsolete -- but we'll still have to support them forever, of course.

Given all this I'd really prefer these functions be placed in a separate module, so that people who don't import the module aren't impacted.

(That said, I'd even more prefer that we just go ahead and implement the native API... it'd be pretty easy.)

Comment (elithrar) [reply src/cloudflare/workers.ts:16]: What’s the path forward? Do we keep telling customers to copy paste JWT validation code from the docs? 

Are JWT native APIs landing this week? Next year?

Comment (kentonv) [reply src/cloudflare/workers.ts:16]: To be clear I was only asking that you move the function to a different module name, like `cloudflare:access` or something. (Though admittedly I don't quite understand why a built-in is better than an npm module here, but I won't fight on that point.)

The native API would be where `ctx.accessCredentials` is just automatically filled in with the verified credentials. I could (prompt Claude to) implement this in an hour probably except that the runtime doesn't know the worker's access team name (to fetch the keys) no audience. Seems like these should be part of the workers config and delivered via the control plane. If someone could spec that out and get the control plane to deliver that info we could ship it next week. But PMs and EMs control priorities here, not me.

But again, assuming we can't prioritize that now, I'm fine with the library, it just has to be a different module name.

Comment (jroyal) [reply src/cloudflare/workers.ts:16]: fwiw I have an engineer about to start working on this from the access side and will probably reach out to you @kentonv soon.

**Comment by elithrar** [comment]
Closing this as it sounds like we're taking a different approach.
